external-crates/move/crates/move-compiler/src/typing/translate.rs

```diff
@@ -2938,12 +2952,101 @@ fn module_call_impl(
     (call, return_)
 }
 
+/// If the constant that we are referencing has an `error` attribute, we need to change the type of
+/// the constant to a u64 since this will be compiled into a u64 error code.
+fn annotated_error_const(context: &mut Context, e: &mut T::Exp, abort_or_assert_str: &str) {
+    let u64_type = Type_::u64(e.ty.loc);
+
+    let mut const_name = None;
+
+    if let sp!(
+        const_loc,
+        T::UnannotatedExp_::Constant(module_ident, constant_name)
+    ) = &mut e.exp
+    {
+        let ConstantInfo {
+            attributes,
+            defined_loc,
+            signature: _,
+        } = context.constant_info(module_ident, constant_name);
+
+        const_name = Some((*defined_loc, *constant_name));
+
+        let has_error_annotation =
+            attributes.contains_key_(&known_attributes::ErrorAttribute.into());
+
+        if has_error_annotation {
+            context.env.check_feature(
+                context.current_package(),
+                FeatureGate::CleverAssertions,
+                *const_loc,
+            );
+
+            *e = T::exp(
+                u64_type.clone(),
+                sp(
+                    *const_loc,
+                    T::UnannotatedExp_::ErrorConstant(Some(*constant_name)),
+                ),
+            );
+        }
+    }
+
+    // println!("Error type: ");
+    // e.ty.print_verbose();
+    // u64_type.print_verbose();
+
+    // println!("Subst before: ");
+    // context.subst.print_verbose();
+    // // NB: this will change the substitution and cause issues
+    // let is_u64_type = subtype_no_report(context, e.ty.clone(), u64_type).is_ok();
```

Comment:
@tnowacki This is where the subst in the context is changing, which is then removing the tvar for the conditional expression.